File: android/app/src/test/java/com/fitlog/app/CatalogRepositoryTest.kt
package com.fitlog.app

import android.content.Context
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import com.fitlog.app.data.CatalogErrorCode
import com.fitlog.app.data.CatalogException
import com.fitlog.app.data.CatalogRepository
import com.fitlog.app.data.CreateCustomExerciseInput
import com.fitlog.app.data.FitLogDatabase
import com.fitlog.app.domain.ExerciseKind
import kotlinx.coroutines.test.runTest
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class CatalogRepositoryTest {

    private lateinit var database: FitLogDatabase
    private lateinit var repository: CatalogRepository

    private val seedJson: String by lazy {
        val context = ApplicationProvider.getApplicationContext<Context>()
        context.assets.open("seed/catalog.json").bufferedReader().use { it.readText() }
    }

    @Before
    fun setUp() {
        database = Room.inMemoryDatabaseBuilder(
            ApplicationProvider.getApplicationContext(),
            FitLogDatabase::class.java,
        ).allowMainThreadQueries().build()

        repository = CatalogRepository(
            dao = database.catalogDao(),
            seedJsonProvider = { seedJson },
            now = { 1_700_000_000_000 },
        )
    }

    @After
    fun tearDown() {
        database.close()
    }

    @Test
    fun `la primera siembra carga el catalogo base`() = runTest {
        val result = repository.ensureSeeded()

        assertTrue(result.seeded)
        assertEquals(12, result.muscleGroups)
        assertEquals(36, result.exercises)

        val snapshot = repository.loadCatalog()
        assertEquals(36, snapshot.exercises.size)
        assertTrue(snapshot.exercises.none { it.isCustom })
    }

    @Test
    fun `las siembras posteriores no duplican filas`() = runTest {
        repository.ensureSeeded()
        val second = repository.ensureSeeded()

        assertFalse(second.seeded)
        assertEquals(36, repository.loadCatalog().exercises.size)
    }

    @Test
    fun `respeta los ejercicios personalizados existentes`() = runTest {
        repository.ensureSeeded()
        repository.createCustomExercise(
            CreateCustomExerciseInput(
                name = "Remo en máquina",
                muscleGroupId = repository.loadCatalog().groups.first().id,
                equipment = "maquina",
                kind = ExerciseKind.STRENGTH,
            )
        )

        val freshRepository = CatalogRepository(
            dao = database.catalogDao(),
            seedJsonProvider = { seedJson },
        )
        freshRepository.ensureSeeded()

        val snapshot = freshRepository.loadCatalog()
        assertEquals(37, snapshot.exercises.size)
        assertEquals(1, snapshot.exercises.count { it.isCustom })
    }

    @Test
    fun `crea un ejercicio propio con slug derivado`() = runTest {
        val groups = repository.loadCatalog().groups
        val created = repository.createCustomExercise(
            CreateCustomExerciseInput(
                name = "  Remo en máquina ",
                muscleGroupId = groups.first().id,
                equipment = "maquina",
                kind = ExerciseKind.STRENGTH,
            )
        )

        assertEquals("remo-en-maquina", created.slug)
        assertEquals("Remo en máquina", created.name)
        assertTrue(created.isCustom)
    }

    @Test
    fun `rechaza nombres vacios y grupos inexistentes`() = runTest {
        val groupId = repository.loadCatalog().groups.first().id

        val blank = runCatching {
            repository.createCustomExercise(
                CreateCustomExerciseInput("   ", groupId, "banda", ExerciseKind.STRENGTH)
            )
        }.exceptionOrNull()
        assertEquals(CatalogErrorCode.INVALID_INPUT, (blank as CatalogException).code)

        val missingGroup = runCatching {
            repository.createCustomExercise(
                CreateCustomExerciseInput("Ejercicio raro", "01ARYZ6S41TSV4RRFFQ69G5FAV", "banda", ExerciseKind.STRENGTH)
            )
        }.exceptionOrNull()
        assertEquals(CatalogErrorCode.MUSCLE_GROUP_NOT_FOUND, (missingGroup as CatalogException).code)
    }

    @Test
    fun `rechaza slugs duplicados`() = runTest {
        val groups = repository.loadCatalog().groups
        val error = runCatching {
            repository.createCustomExercise(
                CreateCustomExerciseInput("Dominadas", groups.first().id, "peso-corporal", ExerciseKind.STRENGTH)
            )
        }.exceptionOrNull()

        assertEquals(CatalogErrorCode.DUPLICATE_SLUG, (error as CatalogException).code)
    }

    @Test
    fun `elimina con borrado logico solo ejercicios propios`() = runTest {
        val snapshot = repository.loadCatalog()
        val base = snapshot.exercises.first { !it.isCustom }

        val protected = runCatching { repository.deleteCustomExercise(base.id) }.exceptionOrNull()
        assertEquals(CatalogErrorCode.BASE_CATALOG_PROTECTED, (protected as CatalogException).code)

        val custom = repository.createCustomExercise(
            CreateCustomExerciseInput("Ejercicio temporal", snapshot.groups.first().id, "banda", ExerciseKind.STRENGTH)
        )
        repository.deleteCustomExercise(custom.id)

        val after = repository.loadCatalog()
        assertNull(after.exercises.firstOrNull { it.id == custom.id })
        assertEquals(36, after.exercises.size)
    }

    @Test
    fun `falla al eliminar un ejercicio inexistente`() = runTest {
        val error = runCatching {
            repository.deleteCustomExercise("01ARYZ6S41TSV4RRFFQ69G5FAV")
        }.exceptionOrNull()

        assertEquals(CatalogErrorCode.EXERCISE_NOT_FOUND, (error as CatalogException).code)
    }
}

File: web/src/test/catalogRepository.test.ts
import { describe, expect, it } from 'vitest';

import { createFitLogDb, type SqlClient } from '@/db/client';
import { applyMigrations } from '@/db/migrate';
import { migrations } from '@/db/migrations';
import {
  CatalogError,
  createCustomExercise,
  deleteCustomExercise,
  loadCatalog,
} from '@/data/catalog';
import { seedCatalog } from '@/data/catalogSeed';

import catalogSeed from '@shared/seed/catalog.json';

import { createNodeSqliteHarness } from './helpers/nodeSqlite';

function first<T>(items: readonly T[]): T {
  const value = items[0];
  if (value === undefined) throw new Error('lista vacia');
  return value;
}

function createDatabase(): { db: ReturnType<typeof createFitLogDb>; client: SqlClient } {
  const harness = createNodeSqliteHarness();
  applyMigrations(harness.runner, migrations);
  seedCatalog(harness.runner, catalogSeed, () => 1);
  return { db: createFitLogDb(harness.client), client: harness.client };
}

describe('loadCatalog', () => {
  it('devuelve los grupos y ejercicios sembrados', async () => {
    const { db } = createDatabase();
    const snapshot = await loadCatalog(db);

    expect(snapshot.groups.length).toBe(catalogSeed.muscle_groups.length);
    expect(snapshot.exercises.length).toBe(catalogSeed.exercises.length);
    expect(snapshot.exercises.every((exercise) => exercise.isCustom === false)).toBe(true);
  });
});

describe('createCustomExercise', () => {
  it('crea un ejercicio propio con slug derivado y ULID', async () => {
    const { db } = createDatabase();
    const group = first(catalogSeed.muscle_groups);

    const created = await createCustomExercise(
      db,
      { name: '  Remo en máquina ', muscleGroupId: group.id, equipment: 'maquina', kind: 'strength' },
      () => 1700000000000
    );

    expect(created.slug).toBe('remo-en-maquina');
    expect(created.name).toBe('Remo en máquina');
    expect(created.isCustom).toBe(true);
    expect(created.id).toHaveLength(26);

    const snapshot = await loadCatalog(db);
    expect(snapshot.exercises.filter((exercise) => exercise.isCustom)).toHaveLength(1);
  });

  it('rechaza nombres vacios', async () => {
    const { db } = createDatabase();
    await expect(
      createCustomExercise(db, {
        name: '   ',
        muscleGroupId: first(catalogSeed.muscle_groups).id,
        equipment: 'barra',
        kind: 'strength',
      })
    ).rejects.toMatchObject({ code: 'invalid_input' });
  });

  it('rechaza grupos musculares inexistentes', async () => {
    const { db } = createDatabase();
    await expect(
      createCustomExercise(db, {
        name: 'Ejercicio raro',
        muscleGroupId: '01ARYZ6S41TSV4RRFFQ69G5FAV',
        equipment: 'barra',
        kind: 'strength',
      })
    ).rejects.toMatchObject({ code: 'muscle_group_not_found' });
  });

  it('rechaza slugs duplicados contra el catalogo base', async () => {
    const { db } = createDatabase();
    const base = first(catalogSeed.exercises.filter((item) => item.slug === 'dominadas'));
    await expect(
      createCustomExercise(db, {
        name: base.name.toUpperCase(),
        muscleGroupId: base.muscle_group_id,
        equipment: 'peso-corporal',
        kind: 'strength',
      })
    ).rejects.toBeInstanceOf(CatalogError);
  });
});

describe('deleteCustomExercise', () => {
  it('aplica borrado logico a ejercicios propios', async () => {
    const { db } = createDatabase();
    const created = await createCustomExercise(
      db,
      {
        name: 'Ejercicio temporal',
        muscleGroupId: first(catalogSeed.muscle_groups).id,
        equipment: 'banda',
        kind: 'strength',
      },
      () => 1700000000000
    );

    await deleteCustomExercise(db, created.id, () => 1700000001000);

    const snapshot = await loadCatalog(db);
    expect(snapshot.exercises.some((exercise) => exercise.id === created.id)).toBe(false);
  });

  it('protege el catalogo base', async () => {
    const { db } = createDatabase();
    await expect(deleteCustomExercise(db, first(catalogSeed.exercises).id)).rejects.toMatchObject({
      code: 'base_catalog_protected',
    });
  });

  it('falla si el ejercicio no existe', async () => {
    const { db } = createDatabase();
    await expect(
      deleteCustomExercise(db, '01ARYZ6S41TSV4RRFFQ69G5FAV')
    ).rejects.toMatchObject({ code: 'exercise_not_found' });
  });
});
